### DISCUSSION TITLE
Emit all genealogical nodes?

### DISCUSSION THREAD

**Discussion by sgravel** [tskit-dev/tskit#2648] [Ideas]
We are interested in using the general_stat function to compute quantities for specific genealogical individuals in genealogical simulations. For example, what is the total span of all the segments contributed by a genealogical individual to the samples. This does not seem to work out of the box using the general_stat in the mode="nodes" function because (I believe) it does not consider contributions of genealogical individuals which happen via a unary node, or prior to the MRCA.  

We would be happy to work to contribute such a feature. One possible workaround (option 1) would be to set all genealogical individuals as samples, do the genealogical simulation, but then set weights in general_stat to be positive only for the "real" samples. I think this would work, but might be quite inefficient in a large genealogy. 

Option 2 is to write a an option to emit_all_genealogical_nodes in a genealogical simulation, which would require a bit more work but might be worth doing. 

I think we will try option 1 to start, but open to suggestions/told that the approach is wrong.

**Comment by petrelharp** [comment]
Is the problem to do with "how to make sure the information is *in* the tree sequence in the first place" or "how to get the summaries *out* of the tree sequence", or both? (maybe both?)

I wonder if it's the former, becuase portions of ancestry for which a non-sample is unary are not retained; if you want to keep these in then you do need to do something (and what to do depends on the simulator?).

But, *if* all the information you need is actually in the tree sequence then I *think* this does what you want, regardless of unary-ness or MRCAs:
```
ts.sample_count_stat(ts.samples(), lambda x: x, 1, polarised=True, span_normalise=False, strict=False, mode="node")
```

**Reply by LukeAndersonTrocme** [reply]
Will dig a little deeper tomorrow, but I'm getting a weird error when I try to run that: 

```  
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [17], line 1
----> 1 ts.sample_count_stat(ts.samples(), lambda x: x, 1, polarised=True, span_normalise=False, strict=False, mode="node")

File ~/anaconda3/envs/msprime-env/lib/python3.10/site-packages/tskit/trees.py:7161, in TreeSequence.sample_count_stat(self, sample_sets, f, output_dim, windows, polarised, mode, span_normalise, strict)
   7159 # helper function for common case where weights are indicators of sample sets
   7160 for U in sample_sets:
-> 7161     if len(U) != len(set(U)):
   7162         raise ValueError(
   7163             "Elements of sample_sets must be lists without repeated elements."
   7164         )
   7165     if len(U) == 0:

TypeError: object of type 'numpy.int32' has no len()
```


edit: solved issue with the following

```
ts.sample_count_stat([ts.samples().tolist()], lambda x: x, 1, polarised=True, span_normalise=False, strict=False, mode="node")
```

**Reply by petrelharp** [reply]
ah, silly mistake on my part

**Comment by sgravel** [comment]
It's true that this is more a question for the simulator than for tskit, since it is about how to record nodes rather than how to parse them. So just to confirm here: if we modified the simulator (here msprime) to emit all nodes within the genealogy, we would be able to use general_stat as discussed to compute the expected total contributions by any individual in the genealogy (including if they do not have a coalescence)? 

There is a second type of summary statistic that we were interested in computing and that might still be doable with  modification of general_stat: it is the amount of pairwise coalescence that occurs in a given genealogical individual (as a generalization to the instantaneous coalescence rate to individuals). In principle, we could do this by computing all IBD and modify the identitysegment object to record the node id where the IBD occurred. But I also think that this could be computed from a modification of the general_stat function where the function f applied not to the weights of a node itself, but to the weights of the children of the node (i.e., if we select a pair of samples (s_1,s_2), the rate of pairwise coalescence in node u is the sum over all pairs (i,j) of children of u of the probability that s_1 descends from node i and s_2 descends from node j, plus the opposite pairing. In other words, the rate of pairwise coalescence in a node u corresponding to a genealogical individual is proportional to \sum_{children i,j of u }(w_i/n_samples w_j/n_samples).  

Does that make sense?

**Reply by nspope** [reply]
The second summary statistic sounds like it could be computed via `TreeSequence.coalescence_time_distribution` (though that's a work in progress and the implementation in the github head of tskit is not efficient). In particular, the general recursive update scheme implemented in #2621 should allow the sort of "combinatorial summing" of children weights up the tree that you describe. 

For example,
```python
import msprime

ts = msprime.sim_ancestry(
    samples=10, population_size=1e4, sequence_length=1e6, recombination_rate=1e-8, random_seed=1024
)

sample_pair = [9,10]

coaltime = ts.coalescence_time_distribution(
    sample_sets=[sample_pair],
    weight_func="pair_coalescence_events",
    span_normalise=True,
)

# this builds a table where each row corresponds to a unique node that has at least one coalescence
# event between the samples in `sample_pair`.
print(coaltime.tables[0].num_records - 1) # number of nodes that are ancestral to `sample_pair`
print(coaltime.tables[0].time) # age of nodes
print(coaltime.tables[0].weights) # span-weighted number of coalescence events occurring in each node   
```
Although it sounds like this would need to be modified this to store node IDs (rather than just ages/weights), so the weights could be mapped back onto the tree sequence.

**Reply by petrelharp** [reply]
Yes, I agree, @sgravel! The class of statistics you outline is also a big and useful one. And, I think slightly easier for implementation, instead of having "weights at the children of each node" we could have "weights for the parent and child of each edge"? (then for instance, if the number of samples below the child of the edge is `x` and below the parent of the edge is `y` then the number of coalescences in the parent of lineages passing through the edge is `x (y-x)`; we can sum this across edges below a node and divide by two to get all the coalescences)

Me and @nspope thought about implementing the coalescence time stats tihs way, but I'm not sure about the status of this idea?

**Reply by nspope** [reply]
I thought about implementing the coalescence event count updates via edges (rather than collections of children), but it seemed that doing so would make specifying "custom" weighting functions kind of cumbersome. For example, to calculate counts of trio coalescence events on a given tree, you'd first need to propagate sample counts up the tree, then propagate pair coalescence events up the tree, then finally calculate trio coalescence events using singles/pairs for the parent & child at a given edge. So, how would this be specified via the "custom weight" interface -- as a tuple of update functions, that are separately passed up the tree in order? 

Whereas if the update acts on all children at once (like @sgravel suggests, and is done in #2621), then only a single update function needs to be specified. It seems like the tradeoff is between making it easier for users to specify custom weighting functions, vs making the internal tree traversal logic simpler. I lean towards the former, but could see it working either way.

**Reply by petrelharp** [reply]
Hm, I remember discussing this and so am probably missing something. But, I don't see why we can't count the trio coalescences with this information? Suppose for each node `u` we know `n(u, A) ` = number of samples in `A` below `u`. Then for sample sets `A`, `B`, and `C`, and an edge with child `c` and parent `p`,
```
n(c, A) * n(c, B) * (n(p, C) - n(c, C))
```
is the number of trees of the form `((xa, xb), xc)` for which the root node is node `p` and the MRCA of `xa` and `xb` are at or below `c`, where `xa` is a sample in `A`, etc. So to get the trio coalescences of the older type for `p` you'd sum this over the edges for which `p` is the parent.
To get the other type of trio coalescence, you'd do:
```
n(c, A) * (n(p, B) - n(c, B)) * (|C| - n(p, C))
```
which is the number of trees of the form `((xa, xb), xc)` for which the MRCA of `xa` and `xb` is `p` (and the root is older).

I haven't checked what these do if there's polytomies. But, hopefully they'd provide a useful definition for that case?

**Reply by sgravel** [reply]
I have not spent as much time as you thinking about this, but from my perspective accessing the weights of children seems much more intuitive, and also a bit more general. For example, if we were to compute the rate of three-way coalescence (simultaneous merger of three lineages in the same individual), I don't think this would be doable with a single sum over edge values. So I think it is a tradeoff between what you think tskit users want, and how much extra cost is involved in coding access to all children nodes.

**Reply by petrelharp** [reply]
That's where we're at, too - me and @nspope are going to discuss the tradeoffs soon. =)

But, if you have concrete examples of things that you really want to do that are not do-able based on that sort of edge-wide information, let us know!

**Reply by General-Solution** [reply]
Hi all, I've started on some Python implementations for the sum over edges statistic:
https://github.com/General-Solution/tskit/blob/main/python/tests/test_parent_child_stats.py

The node_parent_child_stat() methods all implement it slightly differently. Right now node_parent_child_stat_4() is the fastest but might not be doable in C. node_parent_child_stat_2() is slightly slower.

I've also written some tests and benchmarking in the same file.

**Reply by nspope** [reply]
Cool @General-Solution -- you might take a look at the implementation at https://github.com/tskit-dev/tskit/blob/c12f384c6c7820f783a59537b6a31ba3e6a141fb/python/tskit/stats.py#L222 which does a similar sort of thing (to count coalescent events of various sorts -- e.g. number pairwise coalescence events between populations at each node).

**Reply by sgravel** [reply]
Thanks! There does seem to be sizeable overlap, so it might make sense to think about how best to package this. From my perspective, I would imagine that the computation of coalescences per node might be a distinct function, which we can then use to compute a TMRCA distribution, (as you did), but also other ways of picking apart the coalescences (e.g., by geographical area, etc).

**Reply by nspope** [reply]
Totally. Just to be clear, the idea underlying that class is to have a general interface for counting coalescence events of various "topologies" per node (for example, coalescence events of a particular pair of samples, or of a trio of populations). These are stored in a time-sorted table, so that it just takes a binary search to compute the number of events in an arbitrary time window (and calculate things like instantaneous coalescence rate within that window). The class methods are intended to give efficient access to that table wrt time windows, but the node-level values are probably interesting in their own right (e.g. the "weights" could be mapped back onto nodes).

The update scheme implemented in that class is (I think) a pretty good tradeoff between efficient and general, but it's definitely not the only possibility. I completely agree that it's worth hashing out other applications for these types of computations, to figure out what the C-level machinery should ultimately look like.

**Comment by petrelharp** [comment]
> It's true that this is more a question for the simulator than for tskit, since it is about how to record nodes rather than how to parse them. So just to confirm here: if we modified the simulator (here msprime) to emit all nodes within the genealogy, we would be able to use general_stat as discussed to compute the expected total contributions by any individual in the genealogy (including if they do not have a coalescence)?

I don't see why not, yes? Although it's possible I'm missing some oddity, as I haven't done any verification.

Note that keeping this information is pretty straight forward in SLiM (and you can simulate in pedigrees there now).

**Reply by petrelharp** [reply]
Ah, and in msprime, you'd mark everyone as samples in the pedigree.

**Reply by sgravel** [reply]
Right, this would be a way to do this out of the box. Although with a genealogy of 5 million people, this would be a bit heavy (although probably doable, especially if we stop the simulation at the top of the genealogy).

**Comment by jeromekelleher** [comment]
Separate from the the discussion about the stats (which sounds useful and interesting!) is the issue about keeping track of more nodes in msprime.  This has also cropped up in the context of trying to improve performance in general in msprime (https://github.com/tskit-dev/msprime/pull/2121) so maybe addressing the "keeping more of the ARG" in the pedigree simulation would be a good place to start?

cc @GertjanBisschop

**Reply by sgravel** [reply]
It would be a good start, although in our case we will need two kinds of nodes not in the ARG: unary nodes without coalescence nor recombination, and nodes above the MRCA that may be relevant for some genealogical quantities.

**Reply by jeromekelleher** [reply]
Ah, you want the full path through all pedigree nodes. That should be straightforward to do - it's really a case of not doing some logic where we record nodes if there's a coalescence. Ping me offline if you're keen to get it done quickly, we could see what can be done.
